File: projects/CSharp/Benchmarks/SendReceive.cs
using BenchmarkDotNet.Attributes;

using com.chronoxor.proto;
using com.chronoxor.proto.FBE;

namespace Benchmarks
{
    public class MySender1 : Sender, ISenderListener
    {
        private long Size { get; set; }
        private long LogSize { get; set; }

        public long OnSend(byte[] buffer, long offset, long size) { Size += size; return size; }
        public void OnSendLog(string message) { LogSize += message.Length; }
    }

    public class MySender2 : Sender, ISenderListener
    {
        private long Size { get; set; }
        private long LogSize { get; set; }

        public long OnSend(byte[] buffer, long offset, long size) { Size += size; return 0; }
        public void OnSendLog(string message) { LogSize += message.Length; }
    }

    public class MyReceiver : Receiver, IReceiverListener
    {
        private long LogSize { get; set; }

        public void OnReceive(OrderMessage value) {}
        public void OnReceive(BalanceMessage value) {}
        public void OnReceive(AccountMessage value) {}

        public void OnReceiveLog(string message) { LogSize += message.Length; }
    }

    public class SendReceive
    {
        private readonly AccountMessage _account;
        private readonly MySender1 _sender1;
        private readonly MySender2 _sender2;
        private readonly MyReceiver _receiver;

        public SendReceive()
        {
            // Create a new account with some orders
            _account = new AccountMessage(Account.Default);
            _account.body.id = 1;
            _account.body.name = "Test";
            _account.body.state = State.good;
            _account.body.wallet.currency = "USD";
            _account.body.wallet.amount = 1000.0;
            _account.body.asset = new Balance("EUR", 100.0);
            _account.body.orders.Add(new Order(1, "EURUSD", OrderSide.buy, OrderType.market, 1.23456, 1000.0));
            _account.body.orders.Add(new Order(2, "EURUSD", OrderSide.sell, OrderType.limit, 1.0, 100.0));
            _account.body.orders.Add(new Order(3, "EURUSD", OrderSide.buy, OrderType.stop, 1.5, 10.0));

            _sender1 = new MySender1();
            _sender1.Send(_account);

            _sender2 = new MySender2();
            _sender2.Send(_account);

            _receiver = new MyReceiver();
            _receiver.Receive(_sender2.Buffer);
        }

        [Benchmark]
        public void Send()
        {
            // Serialize and send the account
            _sender1.Send(_account);
        }
        /*
        [Benchmark]
        public void SendWithLogs()
        {
            // Enable logging
            _sender1.Logging = true;

            // Serialize and send the account
            _sender1.Send(_account);
        }
        */
        [Benchmark]
        public void Receive()
        {
            // Receive the account from the sender
            _receiver.Receive(_sender2.Buffer);
        }
        /*
        [Benchmark]
        public void ReceiveWithLogs()
        {
            // Enable logging
            _receiver.Logging = true;

            // Receive the account from the sender
            _receiver.Receive(_sender2.Buffer);
        }
        */
    }
}

File: projects/Swift/MyLibrary/Tests/Examples/SendReceive.swift
//
//  SendReceive.swift
//

import XCTest
import proto

fileprivate class MySender: proto.Sender {
    override func onSend(buffer: Data, offset: Int, size: Int) throws -> Int {
        // Send nothing...
        return 0
    }
    
    func onSendLog(message: String) {
        print("onSend: " + message)
    }
}

fileprivate class MyReceiver: proto.Receiver, proto.ReceiverListener {
    func onReceiveLog(message: String) {
        print("onReceive: " + message)
    }
}

class SendReceive: XCTestCase {
    func testSendReceive() {
        let sender = MySender()
        
        // Enable logging
        sender.logging = true
        
        // Create and send a new order
        let order = Order(id: 1, symbol: "EURUSD", side: OrderSide.buy, type: OrderType.market, price: 1.23456, volume: 1000.0)
        _ = try? sender.send(obj: order)
        
        // Create and send a new balance wallet
        let balance = Balance(currency: "USD", amount: 1000.0)
        _ = try? sender.send(obj: balance)
        
        // Create and send a new account with some orders
        let account = Account(id: 1, name: "Test", state: State.good, wallet: Balance(currency: "USD", amount: 1000.0), asset: Balance(currency: "EUR", amount: 100.0), orders: [])
        account.orders.append(Order(id: 1, symbol: "EURUSD", side: OrderSide.buy, type: OrderType.market, price: 1.23456, volume: 1000.0))
        account.orders.append(Order(id: 2, symbol: "EURUSD", side: OrderSide.sell, type: OrderType.limit, price: 1.0, volume: 100.0))
        account.orders.append(Order(id: 3, symbol: "EURUSD", side: OrderSide.buy, type: OrderType.stop, price: 1.5, volume: 10.0))
        _ = try? sender.send(obj: account)
        
        let receiver = MyReceiver()
        
        // Enable logging
        receiver.logging = true
        
        // Receive data from the sender
        try? receiver.receive(buffer: sender.buffer)
    }
}
